File: JavaFundamentals/JavaAdvanced/LinearDataStructures_Ex/MaximumElement.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class MaximumElement {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

        ArrayDeque<Integer> stack = new ArrayDeque<>();
        ArrayDeque<Integer> maxStack = new ArrayDeque<>();

        int N = Integer.parseInt(reader.readLine());
        int max = Integer.MIN_VALUE;
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < N; i++) {
            String[] splitted = reader.readLine().split(" ");
            int type = Integer.parseInt(splitted[0]);
            switch (type){
                case 1:
                    int num = Integer.parseInt(splitted[1]);
                    stack.push(num);
                    if (num > max){
                        max = num;
                        maxStack.push(num);
                    }
                    break;
                case 2:
                    int popped = stack.pop();
                    if (popped == max){
                        maxStack.pop();
                        if (maxStack.isEmpty()){
                            max = Integer.MIN_VALUE;
                        } else {
                            max = maxStack.peek();
                        }
                    }
                    break;
                case 3:
                    sb.append(max).append(System.lineSeparator());
                    break;
            }
        }
        System.out.println(sb);
    }
}

File: CSharpFundamentals/CSharpAdvanced/StacksAndQueues_Ex/MaximumElement.cs
using System;
using System.Collections.Generic;

namespace MaximumElement
{
    class MaximumElement
    {
        private static Stack<int> stack = new Stack<int>();
        private static Stack<int> maxStack = new Stack<int>();
        private static int max;

        static void Main(string[] args)
        {
            max = int.MinValue;
            int N = int.Parse(Console.ReadLine());
            for (int i = 0; i < N; i++)
            {
                string[] command = Console.ReadLine().Split();
                switch (int.Parse(command[0]))
                {
                    case 1:
                        int num = int.Parse(command[1]);
                        stack.Push(num);
                        if (num > max)
                        {
                            max = num;
                            maxStack.Push(num);
                        }
                        break;
                    case 2:
                        int popped = stack.Pop();
                        if (popped == max)
                        {
                            maxStack.Pop();
                            if (maxStack.Count == 0)
                            {
                                max = int.MinValue;
                            }
                            else
                            {
                                max = maxStack.Peek();
                            }
                        }
                        break;
                    case 3:
                        Console.WriteLine(max);
                        break;
                }
            }
        }
    }
}
